Nodejs/Product/Npm/SPI/NodeModules.cs

```diff
@@ -14,47 +14,129 @@
-                        AddModule(new Package(parent, moduleDir, showMissingDevOptionalSubPackages));
-                    }
+        private Dictionary<string, int> _allModules;
+
+        public NodeModules(IRootPackage parent, bool showMissingDevOptionalSubPackages, Dictionary<string, int> allModules = null, int depth = 0) {
+            var modulesBase = Path.Combine(parent.Path, NodejsConstants.NodeModulesFolder);
+
+            _allModules = allModules;
+
+            // This is the first time NodeModules is being created.
+            // Iterate through directories to add everything that's known to be top-level.
+            if (_allModules == null) {
+                _allModules = new Dictionary<string, int>();
+
+                IEnumerable<string> directories = null;
+                try {
+                    directories = Directory.EnumerateDirectories(modulesBase);
+                } catch (IOException) {
+                    // We want to handle DirectoryNotFound, DriveNotFound, PathTooLong
+                } catch (UnauthorizedAccessException) {
                 }
-            }
 
-            var parentPackageJson = parent.PackageJson;
-            if (null != parentPackageJson) {
-                foreach (var dependency in parentPackageJson.AllDependencies) {
-                    Package module = null;
-                    if (!Contains(dependency.Name)) {
-                        var dependencyPath = Path.Combine(modulesBase, dependency.Name);
-                        if (dependencyPath.Length < NativeMethods.MAX_FOLDER_PATH) {
-                            module = new Package(
-                                parent,
-                                dependencyPath,
-                                showMissingDevOptionalSubPackages);
-                            if (parent as IPackage == null || !module.IsMissing || showMissingDevOptionalSubPackages) {
-                                AddModule(module);
+                if (directories != null) {
+                    var bin = string.Format("{0}.bin", Path.DirectorySeparatorChar);
+                    var staging = string.Format("{0}.staging", Path.DirectorySeparatorChar);
+
+                    // Go through every directory in node_modules, and see if it's required as a top-level dependency
+                    // We can check this by looking at whether it begins with a hash tag. 
+                    foreach (var moduleDir in directories) {
+                        if (moduleDir.Length < NativeMethods.MAX_FOLDER_PATH && !moduleDir.EndsWith(bin) && !moduleDir.EndsWith(staging)) {
+                            var packageJson = PackageJsonFactory.Create(new DirectoryPackageJsonSource(moduleDir));
+
+                            if (packageJson != null) {
+                                var enumerator = packageJson.RequiredBy.GetEnumerator();
+                                if (packageJson.RequiredBy.Count() > 0) {

```

Comment:
Could this or the line before throw an exception if the requiredby field is not present (i.e. on NPM < v3)?